File: backend/src/functions/calculate_staff.ts
import dayjs, {Dayjs} from 'dayjs';
import customParseFormat from 'dayjs/plugin/customParseFormat';
import 'dayjs/locale/ru';
import {EquipmentItem, Period} from 'src/types/equipment';

// Импорт данных
const nonOfficerData: EquipmentItem[] = require('../data/non_officer_data.json');
const officerData: EquipmentItem[] = require('../data/officer_data.json');
const womanNonOfficerData: EquipmentItem[] = require('../data/woman_non_officer_data.json');
const womanOfficerData: EquipmentItem[] = require('../data/woman_officer_data.json');

// Инициализация dayjs
dayjs.extend(customParseFormat);
dayjs.locale('ru');

/**
 * Преобразует строку даты в объект Dayjs
 */
function parseDate(date: string | Date): Dayjs {
    if (date instanceof Date) {
        return dayjs(date);
    }
    const parsedDate = dayjs(date, 'DD.MM.YYYY');
    if (!parsedDate.isValid()) {
        console.error('Некорректная дата:', date);
        throw new Error(`Некорректная дата: ${date}`);
    }
    return parsedDate;
}

/**
 * Форматирует дату в строку
 */
function formatDate(date: Dayjs | Date): string {
    return dayjs(date).format('DD.MM.YYYY');
}

/**
 * Находит применимый период для даты выдачи
 * @param item - Предмет обмундирования
 * @param issueDate - Дата выдачи
 * @returns Период выдачи или null, если период не найден
 */
function findApplicablePeriod(
    item: EquipmentItem | undefined,
    issueDate: Dayjs | Date
): Period | null {
    if (!item) return null;

    console.log('Периоды предмета:', item.periods);
    console.log('Поиск периода для даты:', formatDate(issueDate), 'в предмете:', item.item_name);

    const period = item.periods.find((period) => {
        const startDate = period.start_date ? parseDate(period.start_date) : dayjs('1970-01-01');
        const endDate = period.end_date ? parseDate(period.end_date) : dayjs('2100-01-01');
        const issueDateTime = dayjs(issueDate);

        const isInPeriod =
            (issueDateTime.isAfter(startDate) || issueDateTime.isSame(startDate)) &&
            (issueDateTime.isBefore(endDate) || issueDateTime.isSame(endDate));

        console.log('Проверка периода:', {
            periodStart: formatDate(startDate),
            periodEnd: formatDate(endDate),
            isInPeriod,
        });

        return isInPeriod;
    });

    console.log('Найденный период:', period);
    return period;
}

export function calculateEquipmentDates(
    startWorkDate: string,
    officerPromotionDate: string,
    itemName: string,
    gender: 'male' | 'female',
    maternityLeaveStart: string | null = null,
    maternityLeaveDuration: number = 0
) {
    console.log('Начало расчета выдачи для:', {
        startWorkDate,
        officerPromotionDate,
        itemName,
        gender,
    });

    const workStartDate = parseDate(startWorkDate);
    const promotionDate = parseDate(officerPromotionDate);
    const maternityStart = maternityLeaveStart ? parseDate(maternityLeaveStart) : null;
    const currentDate = dayjs();

    const issuances = [];
    let totalQuantity = 0;

    // Выбор данных в зависимости от пола
    const nonOfficerItem =
        gender === 'female'
            ? womanNonOfficerData.find((item) => item.item_name === itemName)
            : nonOfficerData.find((item) => item.item_name === itemName);

    const officerItem =
        gender === 'female'
            ? womanOfficerData.find((item) => item.item_name === itemName)
            : officerData.find((item) => item.item_name === itemName);

    console.log('Найденные предметы:', {
        nonOfficerItem: nonOfficerItem?.item_name,
        officerItem: officerItem?.item_name,
    });

    if (!nonOfficerItem && !officerItem) {
        return {issuances: [], totalQuantity: 0};
    }

    // Определяем начальную дату выдачи
    let currentIssueDate = determineFirstAvailableDate(
        nonOfficerItem || officerItem,
        workStartDate,
        promotionDate,
        !nonOfficerItem && !!officerItem
    );

    console.log('Начальная дата выдачи:', {
        currentIssueDate: formatDate(currentIssueDate),
        workStartDate: formatDate(workStartDate),
        promotionDate: formatDate(promotionDate),
    });

    let lastIssuanceDate = null;

    while (currentIssueDate.isBefore(currentDate) || currentIssueDate.isSame(currentDate)) {
        const isOfficer =
            currentIssueDate.isAfter(promotionDate) || currentIssueDate.isSame(promotionDate);
        const currentItem = isOfficer ? officerItem : nonOfficerItem;

        console.log('Итерация цикла:', {
            currentIssueDate: formatDate(currentIssueDate),
            isOfficer,
            hasItem: !!currentItem,
            itemName: currentItem?.item_name,
        });

        if (!currentItem) {
            console.log('Прерывание: нет текущего предмета');
            break;
        }

        const applicablePeriod = findApplicablePeriod(currentItem, currentIssueDate);
        console.log('Найденный период:', applicablePeriod);

        if (!applicablePeriod) {
            console.log('Прерывание: не найден применимый период');
            break;
        }

        // Проверка переходного периода
        if (lastIssuanceDate && !isOfficer) {
            const nextIssueDate = dayjs(lastIssuanceDate).add(
                applicablePeriod.period_months,
                'month'
            );

            console.log('Проверка переходного периода:', {
                lastIssueDate: formatDate(lastIssuanceDate),
                nextIssueDate: formatDate(nextIssueDate),
                promotionDate: formatDate(promotionDate),
                isAfterPromotion: nextIssueDate.isAfter(promotionDate),
                currentItem: currentItem.item_name,
                period_months: applicablePeriod.period_months,
            });

            // Если срок следующей выдачи заходит за дату присвоения звания,
            // нужно доносить текущий предмет до конца срока
            if (nextIssueDate.isAfter(promotionDate)) {
                if (nextIssueDate.isAfter(currentDate)) {
                    break;
                }
                // Следующая выдача будет по офицерской норме после окончания срока носки
                currentIssueDate = nextIssueDate;
                continue;
            }
        }

        // Проверка срока предыдущей выдачи
        if (
            lastIssuanceDate &&
            !isPreviousIssuanceExpired(
                lastIssuanceDate,
                applicablePeriod.period_months,
                currentIssueDate,
                maternityStart,
                maternityLeaveDuration
            )
        ) {
            currentIssueDate = dayjs(lastIssuanceDate).add(applicablePeriod.period_months, 'month');
            continue;
        }

        // Добавление выдачи
        issuances.push({
            date: formatDate(currentIssueDate),
            quantity: applicablePeriod.quantity,
            used: 0,
        });
        totalQuantity += applicablePeriod.quantity;

        lastIssuanceDate = currentIssueDate;
        currentIssueDate = currentIssueDate.add(applicablePeriod.period_months, 'month');

        // Учет декретного отпуска
        if (maternityStart && gender === 'female') {
            if (
                currentIssueDate.isBefore(maternityStart) &&
                currentIssueDate
                    .add(applicablePeriod.period_months, 'month')
                    .isAfter(maternityStart)
            ) {
                currentIssueDate = currentIssueDate.add(maternityLeaveDuration, 'month');
            }
        }
    }

    console.log('Результат расчета:', {
        itemName,
        totalQuantity,
        issuancesCount: issuances.length,
    });

    return {
        issuances,
        totalQuantity,
    };
}

/**
 * Обрабатывает все предметы для сотрудника
 */
export function processEquipment(
    startWorkDate: Date | string,
    officerPromotionDate: Date | string,
    gender: 'male' | 'female',
    maternityLeaveStart: Date | string | null = null,
    maternityLeaveDuration: number = 0
) {
    const formattedStartDate = formatDate(dayjs(startWorkDate));
    const formattedOfficerDate = formatDate(dayjs(officerPromotionDate));
    const formattedMaternityDate = maternityLeaveStart
        ? formatDate(dayjs(maternityLeaveStart))
        : null;

    const allItems = new Set([
        ...(gender === 'female' ? womanOfficerData : officerData).map((item) => item.item_name),
        ...(gender === 'female' ? womanNonOfficerData : nonOfficerData).map(
            (item) => item.item_name
        ),
    ]);

    return Array.from(allItems).map((itemName) => {
        const itemResult = calculateEquipmentDates(
            formattedStartDate,
            formattedOfficerDate,
            itemName,
            gender,
            formattedMaternityDate,
            maternityLeaveDuration
        );

        // Получаем информацию о денежной компенсации
        const officerItem = (gender === 'female' ? womanOfficerData : officerData).find(
            (item) => item.item_name === itemName
        );
        const nonOfficerItem = (gender === 'female' ? womanNonOfficerData : nonOfficerData).find(
            (item) => item.item_name === itemName
        );
        const cashValue = (officerItem || nonOfficerItem)?.cash || 0;

        return {
            name: itemName,
            issuances: itemResult.issuances,
            totalQuantity: itemResult.totalQuantity,
            cash: cashValue,
        };
    });
}

/**
 * Определяет дату первой возможной выдачи предмета
 */
function determineFirstAvailableDate(
    item: EquipmentItem,
    workStartDate: Dayjs,
    promotionDate: Dayjs,
    isOfficerOnly: boolean
): Dayjs {
    const firstPeriod = item.periods[0];
    const periodStartDate = firstPeriod.start_date
        ? parseDate(firstPeriod.start_date)
        : dayjs('1970-01-01');

    if (isOfficerOnly) {
        const earliestDate = dayjs(promotionDate).isAfter(periodStartDate)
            ? promotionDate
            : periodStartDate;
        return dayjs(earliestDate).isAfter(workStartDate) ? earliestDate : workStartDate;
    }

    return dayjs(periodStartDate).isAfter(workStartDate) ? periodStartDate : workStartDate;
}

/**
 * Проверяет, истек ли срок предыдущей выдачи
 */
function isPreviousIssuanceExpired(
    lastIssueDate: Dayjs,
    periodMonths: number,
    currentDate: Dayjs,
    maternityStart: Dayjs | null,
    maternityLeaveDuration: number
): boolean {
    if (!lastIssueDate) return true;
    let expirationDate = dayjs(lastIssueDate).add(periodMonths, 'month');

    if (
        maternityStart &&
        dayjs(lastIssueDate).isBefore(maternityStart) &&
        dayjs(maternityStart).isBefore(expirationDate)
    ) {
        expirationDate = expirationDate.add(maternityLeaveDuration, 'month');
    }

    return dayjs(currentDate).isAfter(expirationDate) || dayjs(currentDate).isSame(expirationDate);
}

File: backend_py/app/functions/calculate_staff.py
import json
from datetime import datetime, timedelta
from typing import List, Dict, Optional
from dataclasses import dataclass
import os
from pathlib import Path

@dataclass
class Period:
    start_date: Optional[str]
    end_date: Optional[str]
    period_months: int
    quantity: int

@dataclass
class EquipmentItem:
    item_name: str
    periods: List[Period]
    cash: Optional[float] = None

@dataclass
class Issuance:
    date: str
    quantity: int
    used: int = 0

def load_data() -> tuple[List[EquipmentItem], ...]:
    data_dir = Path(__file__).parent.parent / 'data'
    
    def load_json(filename: str) -> List[EquipmentItem]:
        with open(data_dir / filename, 'r', encoding='utf-8') as f:
            data = json.load(f)
            return [
                EquipmentItem(
                    item_name=item['item_name'],
                    periods=[
                        Period(
                            start_date=period.get('start_date'),
                            end_date=period.get('end_date'),
                            period_months=period['period_months'],
                            quantity=period['quantity']
                        ) 
                        for period in item['periods']
                    ],
                    cash=item.get('cash')
                )
                for item in data
            ]

    return (
        load_json('non_officer_data.json'),
        load_json('officer_data.json'),
        load_json('woman_non_officer_data.json'),
        load_json('woman_officer_data.json')
    )

def parse_date(date_str: str) -> datetime:
    try:
        return datetime.strptime(date_str, '%d.%m.%Y')
    except ValueError as e:
        raise ValueError(f"Invalid date format: {date_str}") from e

def format_date(date: datetime) -> str:
    return date.strftime('%d.%m.%Y')

def find_applicable_period(item: EquipmentItem, issue_date: datetime) -> Optional[Period]:
    if not item:
        return None

    for period in item.periods:
        start_date = parse_date(period.start_date) if period.start_date else datetime(1970, 1, 1)
        end_date = parse_date(period.end_date) if period.end_date else datetime(2100, 1, 1)
        
        if start_date <= issue_date <= end_date:
            return period
    
    return None

def calculate_equipment_dates(
    start_work_date: str,
    officer_promotion_date: str,
    item_name: str,
    gender: str,
    maternity_leave_start: Optional[str] = None,
    maternity_leave_duration: int = 0
) -> Dict:
    non_officer_data, officer_data, woman_non_officer_data, woman_officer_data = load_data()
    
    work_start = parse_date(start_work_date)
    promotion_date = parse_date(officer_promotion_date)
    maternity_start = parse_date(maternity_leave_start) if maternity_leave_start else None
    current_date = datetime.now()

    # Select data based on gender
    non_officer_item = next(
        (item for item in (woman_non_officer_data if gender == 'female' else non_officer_data)
         if item.item_name == item_name),
        None
    )
    officer_item = next(
        (item for item in (woman_officer_data if gender == 'female' else officer_data)
         if item.item_name == item_name),
        None
    )

    if not non_officer_item and not officer_item:
        return {'issuances': [], 'total_quantity': 0}

    issuances = []
    total_quantity = 0
    current_issue_date = work_start
    last_issuance_date = None

    while current_issue_date <= current_date:
        is_officer = current_issue_date >= promotion_date
        current_item = officer_item if is_officer else non_officer_item

        if not current_item:
            break

        applicable_period = find_applicable_period(current_item, current_issue_date)
        if not applicable_period:
            break

        # Handle transition period
        if last_issuance_date and not is_officer:
            next_issue_date = last_issuance_date + timedelta(days=applicable_period.period_months * 30)
            if next_issue_date > promotion_date:
                current_issue_date = next_issue_date
                continue

        # Check previous issuance expiration
        if last_issuance_date:
            expiration_date = last_issuance_date + timedelta(days=applicable_period.period_months * 30)
            if maternity_start and gender == 'female':
                if last_issuance_date < maternity_start < expiration_date:
                    expiration_date += timedelta(days=maternity_leave_duration * 30)
            
            if current_issue_date < expiration_date:
                current_issue_date = expiration_date
                continue

        issuances.append({
            'date': format_date(current_issue_date),
            'quantity': applicable_period.quantity,
            'used': 0
        })
        total_quantity += applicable_period.quantity

        last_issuance_date = current_issue_date
        current_issue_date += timedelta(days=applicable_period.period_months * 30)

        # Account for maternity leave
        if maternity_start and gender == 'female':
            if current_issue_date < maternity_start < (current_issue_date + timedelta(days=applicable_period.period_months * 30)):
                current_issue_date += timedelta(days=maternity_leave_duration * 30)

    return {
        'issuances': issuances,
        'total_quantity': total_quantity
    }

def process_equipment(
    start_work_date: str,
    officer_promotion_date: str,
    gender: str,
    maternity_leave_start: Optional[str] = None,
    maternity_leave_duration: int = 0
) -> List[Dict]:
    non_officer_data, officer_data, woman_non_officer_data, woman_officer_data = load_data()
    
    # Get all unique item names
    all_items = set()
    if gender == 'female':
        all_items.update(item.item_name for item in woman_officer_data)
        all_items.update(item.item_name for item in woman_non_officer_data)
    else:
        all_items.update(item.item_name for item in officer_data)
        all_items.update(item.item_name for item in non_officer_data)

    results = []
    for item_name in all_items:
        item_result = calculate_equipment_dates(
            start_work_date,
            officer_promotion_date,
            item_name,
            gender,
            maternity_leave_start,
            maternity_leave_duration
        )
        
        # Get cash compensation info
        data_source = woman_officer_data if gender == 'female' else officer_data
        officer_item = next((item for item in data_source if item.item_name == item_name), None)
        data_source = woman_non_officer_data if gender == 'female' else non_officer_data
        non_officer_item = next((item for item in data_source if item.item_name == item_name), None)
        
        cash_value = (officer_item or non_officer_item).cash if (officer_item or non_officer_item) else 0

        results.append({
            'name': item_name,
            'issuances': item_result['issuances'],
            'total_quantity': item_result['total_quantity'],
            'cash': cash_value
        })

    return results